langchain_tools_demo/tools.py

```diff
@@ -36,7 +37,17 @@ def get_id_token():
         CREDENTIALS, _ = google.auth.default()
     if not CREDENTIALS.valid:
         CREDENTIALS.refresh(Request())
-    return CREDENTIALS.id_token
+    if hasattr(CREDENTIALS, "id_token"):
+        return CREDENTIALS.id_token
+    else:
+        # Use Compute Engine default credential
+        CREDENTIALS = compute_engine.IDTokenCredentials(
+            request=Request(),
+            target_audience=BASE_URL,
+            use_metadata_identity_endpoint=True,
+        )
+        CREDENTIALS.refresh(Request())
+        return CREDENTIALS.token
```

Comment:
I think there is a small bug on refresh:

Consider the following flow:

1. get_id_token is called 
2. CREDENTIALS are set to google.auth.default()
3. CREDENTIALS doesn't have id_token
4. CREDENTIALS is replaced w/ IDToken credentials 
5. CREDENTIALS.token is returned
6.  -- some time passes --
7. get_id_token is called 
8. CREDENTIALS doesn't have id_token
9. CREDENTIALS is replaced w/ IDToken credentials

Reply:
modified